discovery.go

```diff
@@ -0,0 +1,225 @@
+					d.doneDiscovery <- topic
+				}()
+			}
+		case topic := <-d.doneDiscovery:
+			delete(d.ongoingDiscovery, topic)
+		case <-d.p.ctx.Done():
+			log.Info("pubsub discoverloop shutting down")
+			return
+		}
+	}
+}
+
+// advertise advertises this node's interest in a topic to a discovery service
+func (d *discover) advertise(topic string) {
+	if d.discovery != nil {
+		advertisingCtx, cancel := context.WithCancel(d.p.ctx)
+
+		if _, ok := d.advertising[topic]; ok {
+			return
+		}
+		d.advertising[topic] = cancel
+
+		go func() {
+			next, err := d.discovery.Advertise(advertisingCtx, topic)
+			if err != nil {
+				log.Warningf("bootstrap: error providing rendezvous for %s: %s", topic, err.Error())
+			}
+
+			t := time.NewTimer(next)
+			for {
+				select {
+				case <-t.C:
+					next, err = d.discovery.Advertise(advertisingCtx, topic)
+					if err != nil {
+						log.Warningf("bootstrap: error providing rendezvous for %s: %s", topic, err.Error())
+					}
+					t.Reset(next)
+				case <-advertisingCtx.Done():
+					t.Stop()
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (d *discover) stopAdvertise(topic string) {
+	if advertiseCancel, ok := d.advertising[topic]; ok {
+		advertiseCancel()
+		delete(d.advertising, topic)
+	}
+}
+
+// bootstrap handles initial bootstrapping of peers for a given topic
+func (d *discover) bootstrap(async bool, topic string, dOpts ...discovery.Option) {
+	newMux := &sync.Mutex{}
+	newMux.Lock()
+	bootstrapMux, bootstrapStarted := d.bootstrapped.LoadOrStore(topic, newMux)
+
+	if bootstrapStarted {
+		newMux.Unlock()
+		if async {
+			return
+		}
+		mux := bootstrapMux.(*sync.Mutex)
+		mux.Lock()
+		mux.Unlock()
+	} else {
+		d.handleDiscovery(d.p.ctx, topic, dOpts)
+		newMux.Unlock()
+	}
+}
+
+func (d *discover) handleDiscovery(ctx context.Context, topic string, opts []discovery.Option) {
+	if d.discovery == nil {
+		return
+	}
+
+	discoverCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	peerChan, err := d.discovery.FindPeers(discoverCtx, topic, opts...)
+	if err != nil {
+		log.Debugf("error finding peers for topic %s: %v", topic, err)
+		return
+	}
+
+	findPeers := make(map[peer.ID]struct{})
+
+loop:
+	for {
+		select {
+		case pi, more := <-peerChan:
+			if pi.ID == d.p.host.ID() || pi.ID == "" {
+				if more {
+					continue
+				} else {
+					break loop
+				}
+			}
+
+			findPeers[pi.ID] = struct{}{}
+
+			go func(pi peer.AddrInfo) {
+				ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+				defer cancel()
+
+				err := d.p.host.Connect(ctx, pi)
+				if err != nil {
+					log.Debugf("Error connecting to pubsub peer %s: %s", pi.ID, err.Error())
+					return
+				}
+			}(pi)
+
+			if !more {
+				break loop
+			}
+
+		case <-discoverCtx.Done():
+			log.Infof("pubsub discovery for %s timeout", topic)
+			return
+		}
+	}
+
+	if len(findPeers) > 0 {
+		req := &reqNewPeerNotifier{initPeers: findPeers, resp: make(chan chan peer.ID, 1)}
```

Comment:
why are you doing this? the event loop will receive connection notifications anyway.

Reply:
This request does two things.
1) I need to check not just inbound peers, but peers I'm already subscribed to. So this allows me to fill the channel with the intersection of peers I care about and peers I'm already connected to. This could be done with more memory usage by either sending all peers instead of the intersection or by trying to utilize the ListPeers method. Using the ListPeers method may be racy and a bit tricky to work around, but if we're ok being not perfectly precise here we can probably figure this out.
2) We need to be notified about incoming messages on a unique channel per topic/bootstrapping goroutine. We could also have a single discover channel that receives these messages and dispatches them to the goroutines. The way it was done seemed more flexible/reusable since it's more about peer-state then discovery, but this could be moved around.

If you'd like this to go away that's fine, just let me know which path you'd like to take.

Reply:
please take this out of the pubsub mainloop.